### DISCUSSION TITLE
babystepping with klipper

### DISCUSSION THREAD

**Discussion by antlestxp** [UnchartedBull/OctoDash#2321] [Ideas]
is it possible to change the code for babystepping? I am using klipper and it doesn't understand M290 commands. It uses "SET_CODE_OFFSET Z_ADJUST" to offset the nozzle. Example "SET_GCODE_OFFSET Z_ADJUST=0.1 MOVE=1" would offset the nozzle 0.1mm for all gcode after it was submitted.

**Comment by jneilliii** [comment]
SSH to the pi and run the command `nano ~/.config/octodash/config.json` and find the section for printer and modify the babystepping command.

**Comment by antlestxp** [comment]
I see that you can change the babystep command in the config file but it
appends the z offset after the code. The syntax is wrong. I would need to
change the config and the step value that is added to the end of the
command.

**Comment by jneilliii** [comment]
I would think you could do that on the klipper side, potentially with a gcode macro script or something to define the M290 command. I thought that was the one of klipper's super-powers.

**Comment by antlestxp** [comment]
Oh I will have to look into that. I just moved over from RRF. Still a bunch
to learn.

**Comment by antlestxp** [comment]
You are on the right track. I can create a macro that would read the m290
command and execute the command that klipper understands. The hurdle is
that octodash outputs for example "m290 z0.01". When I use that Klipper
doesn't like it. I believe the decimal is not allowed to be in the command.
It doesn't like the space either but that is easy to change in the octodash
config.json.

I did this macro
[gcode_macro BABYZ0.01]
#   Babystep Up .1
gcode:
    SET_GCODE_OFFSET Z_ADJUST=0.01 MOVE=1

[gcode_macro BABYZ-0.01]
#   Babystep Down .1
gcode:
    SET_GCODE_OFFSET Z_ADJUST=-0.01 MOVE=1

Klipper freaks out when reading BABYZ0.01. I think if I could get octodash
to output 1 instead of 0.01 everything would work

**Comment by jneilliii** [comment]
Did you try parameters? https://klipper.discourse.group/t/macro-creation-tutorial/30/4

```
[gcode_macro BABYZ]
  SET_GCODE_OFFSET Z_ADJUST={ params.babystep } MOVE=1
```

then set the config.json value to `BABYZ babystep=`.

**Comment by antlestxp** [comment]
This is super helpful. I will have to give it a try after this print is
done.

**Reply by jneilliii** [reply]
make sure you catch my edit on github, what was emailed was slightly wrong....

**Reply by antlestxp** [reply]
caught your edit

**Comment by antlestxp** [comment]
with that code in play I get a different error

Error on 'SET_GCODE_OFFSET Z_ADJUST= MOVE=1': unable to parse

So it macro is understanding what to pickup but fails to inject the "0.01" or whatever octodash is suppose to append to the end of the gcode command

**Reply by jneilliii** [reply]
my bad, looks like I forgot a bit...

```
[gcode_macro BABYZ]
gcode:
	SET_GCODE_OFFSET Z_ADJUST={ params.babystep } MOVE=1
```

**Reply by antlestxp** [reply]
Oh saw that the "gcode:" was missing and added it before I updated my configuration. As it stands it doesn't import what octodash appends at the end of the command.

**Reply by jneilliii** [reply]
Unfortunate, someone more knowledgeable to klipper would have to speak up on this I'm afraid. I've never used it before. They do have a discord though you may want to check out.

**Reply by antlestxp** [reply]
I appreciate all your help. I think you got me a pretty good head start

**Reply by jneilliii** [reply]
so this came up again in the discord and we figured out if you use this command in config.json it will work.

```
SET_GCODE_OFFSET_ADJUST MOVE=1 Z=
```

**Comment by antlestxp** [comment]
I will give it a try when I get home.

**Comment by antlestxp** [comment]
no dice. it doesn't respond and terminal spits out unknown command. my
setup must have something wrong with it. i added "set_gcode_offset_adjust
move=1 z= to the json with no macros in klipper.

Unknown command:"SET_GCODE_OFFSET_ADJUST"

this is what terminal spits out when a zstep is requested from octodash

**Reply by UnchartedBull** [reply]
`SET_GCODE_OFFSET MOVE=1 Z_ADJUST=` works for me as `zBabystepGCode` just in case this is still a problem

**Comment by antlestxp** [comment]
sorry i was out of the country. Is there anything else that needs to be
added to the klipper config or octoprint other than adjusting that line in
the octodash json?

**Comment by Greysqrl** [comment]
It's working for me with the following macro:

[gcode_macro m290]
gcode:
  SET_GCODE_OFFSET Z_ADJUST={params.Z|default(0.0)} MOVE=1

Be careful to indent the "SET_GCODE..." line properly by two spaces

With this a positive Babystep change on Octodash will raise the nozzle vice versa

**Comment by antlestxp** [comment]
It was the spaces!
